veps/sig-network/passt-migration-proposal.md

```diff
@@ -0,0 +1,290 @@
+It is already used in KubeVirt as a network binding plugin, shipped with the virt-launcher container image, and is documented in the [user guide](https://kubevirt.io/user-guide/network/net_binding_plugins/passt/#passt-binding).
+
+### Live Migration with passt
+![Live Migration with passt ](./images/passtMigrationDiagram.png)
+
+### Support live migration
+
+The [Live Migration](https://kubevirt.io/user-guide/compute/live_migration/#live-migration) process triggers a chain of events, which at a high level networking perspective involves the following steps:
+- A target virt-launcher pod is prepared, with applicable network plumbing.
+- Libvirt is executed. It receives the DomainXML data from the source VM and it also launches the `passt` binary, before executing QEMU which starts the VM.
+- Once both source and target VMs are up, depending on [the migration strategy](https://kubevirt.io/user-guide/compute/live_migration/#understanding-different-migration-strategies) a switchover occurs from the source to the target.
+- The passt instance running on the source sets the TCP_REPAIR socket option on TCP sockets, thus freezing the state of the connection and allowing a dump of all the information needed to restore it (sequences, congestion window parameters, pending data queues - IP addresses and ports are already known.).
+- It communicates the state to peer `passt` process running in the target host, via [facilities provided by vhost-user](https://qemu-project.gitlab.io/qemu/interop/vhost-user.html#migrating-back-end-state)
+- Upon receipt, the target `passt` sets the `TCP_REPAIR` socket option on a new target socket, it restores the state, and eventually clears of the `TCP_REPAIR` socket option. 
+- The connection is now seamlessly resumed, continuing from where it left off.
+
+#### Linux capabilities
+As explained above, passt needs to set and clear the `TCP_REPAIR` socket option. Since this operation enables "hijacking" connections, it is implemented in Linux as a privileged operation requiring `CAP_NET_ADMIN`.
+Since passt runs in the unprivileged Linux context of virt-launcher, it lacks the necessary `CAP_NET_ADMIN` capability. Therefore, passt delegates this operation to a privileged helper running in virt-handler.
+
+#### The Role of passt-repair
+
+The helper responsible for handling `TCP_REPAIR` is called [passt-repair](https://man.archlinux.org/man/passt-repair.1.en). 
+It is a [C program]((https://passt.top/passt/tree/passt-repair.c) included in the `passt` package.
+`passt` creates a Unix Domain Socket in a predefined location (which is available to both virt-launcher and virt-handler as a `hostPath` mounted volume).
+It starts a server that binds to the socket, and waits for incoming messages.
+`passt-repair` is executed as a **one-off process**, separately on migration source and migration target.
```

Comment:
Could you please elaborate why `passt-repair` behaves as a client and not as a server? After all, it provides a service for passt.

Reply:
I don't follow what is the reason for this behavior.
An alternative could be for the `passt-repair` program to behave as a server for passt to connect to when it needs.

Reply:
That's because, generally speaking, in KubeVirt, a model where the privileged component **initiates** actions is preferred. Source: past discussions I read/heard. If you need more details please ask.

In any case, that's how passt and passt-repair interact now. They're not part of KubeVirt.

Reply:
An additional reason for this (admittedly somewhat odd-looking) approach is that passt's security self-isolation means that after initial setup it can no longer see the filesystem.  That makes it impossible it for it to connect out to a passt-repair service at migration time.  Having passt be the "server" at the unix socket level means it can hold on to a listening fd when it isolates; the same approach doesn't really work if it's a unix socket client.

Reply:
Thank you for clarifying @sbrivio-rh @dgibson.